framework/src/araui/assets/ui_font_asset.c

```diff
@@ -0,0 +1,280 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <string.h>
+#include <stdint.h>
+#include <sys/types.h>
+#include <araui/ui_commons.h>
+#include <araui/ui_asset.h>
+#include "libs/ui_map.h"
+#include "libs/ui_list.h"
+#include "ui_core_internal.h"
+#include "ui_asset_internal.h"
+#include "ui_commons_internal.h"
+#include "ui_request_callback.h"
+#include "ui_log.h"
+
+#define DEFAULT_GLYPH_MAP_CAPACITY 256
+
+static void _ui_font_asset_destroy_func(void *userdata);
+
+ui_asset_t ui_font_asset_create_from_file(const char *filename)
+{
+	ui_font_asset_body_t *body;
+	FILE *file;
+	uint32_t font_height;
+	uint32_t font_width;
+	uint32_t glyph_num;
+	uint32_t utf8_code;
+	uint32_t data_size;
+	int glyph_idx;
+	char *extension;
+	ui_asset_glyph_t *glyph;
+
+	if (!ui_is_running()) {
+		UI_LOGE("error: UI framework is not running!\n");
+		return UI_NULL;
+	}
+
+	if (!filename) {
+		UI_LOGE("error: invaild parameter!\n");
+		return UI_NULL;
+	}
+
+	extension = strrchr(filename, '.');
+	if (strncmp(extension, ".fnt", 5) != 0) {
+		UI_LOGE("error: unsupported font file!\n");
+		return UI_NULL;
+	}
+
+	// file open
+	file = fopen(filename, "r");
+	if (!file) {
+		UI_LOGE("error: failed to open file!\n");
+		return UI_NULL;
+	}
+
+	// read header of file
+	ui_fread(&font_height, 4, 1, file);
+	ui_fread(&glyph_num, 4, 1, file);
+
+	// make font asset body
+	body = (ui_font_asset_body_t *)UI_ALLOC(sizeof(ui_font_asset_body_t));
+	if (!body) {
+		(void)fclose(file);
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_font_asset_body_t));
+	((ui_asset_body_t *)body)->type = UI_FONT_ASSET;
+	body->height = font_height;
+
+	// read glyph from the buffer
+	ui_map_init(&body->glyph_map, DEFAULT_GLYPH_MAP_CAPACITY);
```

Comment:
The following functions may fail, but don't return any errors and corresponding callers do nothing.
Please, consider all possible error cases.
void ui_map_init(ui_map_t *map, size_t capacity);
void ui_map_deinit(ui_map_t *map);
void ui_map_push(ui_map_t *map, uint32_t key, void *item);
void *ui_map_get(ui_map_t *map, uint32_t key);

Reply:
I'll add return type and the error logs.
This logic can skip the error (only the logging is needed).
Because unadded glyph is not showing.